### PR TITLE
Fix #18059: Width and height of custom window not changeable via script

### LINKED ISSUE #18059: Plugin: width and height of custom window not changeable via script
Problem: A custom window that is not resizable by the user (by dragging the bottom right corner), is also not resizable by the script. Attempting to change the `width` and `height` properties has no effect.

Note: If the window is resizable, everything works as expected.

Cause: When opening a custom window, there are optional parameters `minHeight` and `maxHeight`. If they are specified, the window will be resizable. Otherwise, those values will internally be set to whatever `height` is. In this case, when the window's height is changed by the script, it will be trimmed to fit into the range `[minHeight, maxHeight]`, i.e. the attempted change has no effect. Same for `width`.

Solution: If the window is not resizable, `minHeight` and `maxHeight` should either be adjusted or ignored.

Build: OpenRCT2, v0.4.1-312-gd263464 (d263464 on develop)

### PR DESCRIPTION
**Pull request by Sadret**

### CODE REVIEW

File: distribution/changelog.txt [hunk 1]
```diff
@@ -67,8 +67,9 @@
 - Fix: [#18008] Steeplechase S-bends has multiple gaps visible in the tracks.
 - Fix: [#18009] Visual glitch with litter at edge of sloped path.
 - Fix: [#18026] Park rating drops to 0 with more than 32k guests, total ride excitement or intensity.
-- Fix: [#18051] Visual glitch with Mine Ride's large unbanked turn.
 - Fix: [#18032] All non-interactive widgets (labels, groupboxes) produce sound when clicked.
+- Fix: [#18051] Visual glitch with Mine Ride's large unbanked turn.
```
Comment (Broxzier) [inline distribution/changelog.txt:71]: Whoops, nice catch.

File: distribution/changelog.txt [hunk 2]
```diff
@@ -67,8 +67,9 @@
...
+- Fix: [#18059] Width and height of custom window not changeable via script.
```
Comment (Broxzier) [inline distribution/changelog.txt:72]:
```suggestion
- Fix: [#18059] [Plugin] Width and height of custom window not changeable via script.
```

**Review by Broxzier — APPROVED** [review]

**Comment by IntelOrca** [comment]
I am not sure why this would be a bug. I think internal code has the same restriction. The code should set the min/max if it wants to change it.

**Comment by Sadret** [comment]
> I am not sure why this would be a bug. I think internal code has the same restriction. The code should set the min/max if it wants to change it.

From the viewpoint of a plugin developer that does not know how the implementation: They create a window, specifying width and height, but not minWidth/maxWidth and minHeight/maxHeight. Width/height seems to be writable, so they try to resize the window, which does not work. There is no hint at all that the min/max values are the reason for that.

**Comment by Sadret** [comment]
In particular, the min/max values are optional when creating a window, but are still internally set to the width/height value. This is not transparent at all.

**Comment by Broxzier** [comment]
For plug-in developers this can be really confusing, since they don't set a minimal or maximum size anywhere. Wouldn't it be better to simply set the min to zero and max to 2^31-1 to get around this?

**Comment by Sadret** [comment]
When I created the issue and the PR, I only thought about custom windows and did not realize that it would also effect vanilla windows. So I have to overthink.

The main problem I have with the current state is this: When I create a window using only width and height, I am expecting to be able to change width and height because they are marked in the documentation as mutable. But this is not the case (at least not without the min/max workaround). So, in the current state, everyone who wants to use this part of the API either has to ask or has to look into the implementation, and even then, a workaround should not be necessary.

This is why I submitted the issue and PR. I still think that behaviour-wise the changes provide the best solution for custom windows. Implementation-wise it might be considered unclean/hack though. For regular windows, I think that the behaviour is fine, I do not really see a problem with plugins being able to change the size of non-resizable windows. Is there an argument against it? It will probably never happen anyway, but if a plugin really wants to misuse it, there are already much better ways to provide bad user experience. 

Also, keep in mind that the behaviour of resizable windows is not effected at all.

Otherwise, as @IntelOrca suggested, adding documentation is a valid options. Something like this:
- width: When writing to this property, the value will be clamped to [minWidth,maxWidth]
- minWidth, maxWidth: If unspecified during window creation, they default to width

Same for height, of course. This will work, but I personally think the current behaviour is a bit tedious for plugin devs, so I would rather change the behaviour than the documentation.

Summary of my (subjective) viewpoint:
- If a dev wants to change the size of their custom window, then `window.width = value` should do the trick.
- It is not a problem if plugins can change the size of "non-resizable" windows.

What are your thoughts on this?

**Comment by Sadret** [comment]
> For plug-in developers this can be really confusing, since they don't set a minimal or maximum size anywhere. Wouldn't it be better to simply set the min to zero and max to 2^31-1 to get around this?

The problem with this approach is that a custom window is handled as resizable if the min and max values are different. So we would need to introduce a new boolean `resizable` and use that during window creation, which is definitely also possible wihtout much overhead.

**Comment by IntelOrca** [comment]
Given this behaviour only affects windows that can't resize, I am happy with it. I think the documentation should still be updated to explain how it works.

**Comment by Sadret** [comment]
I added documentation to `Window` and `WindowDesc`. Let me know what you think.
I also increased the api version.

**Review by IntelOrca — APPROVED** [review]

**Pull request merged**